pkg/ingester/wal.go

```diff
@@ -569,6 +622,35 @@ type samplesWithUserID struct {
 	userID  string
 }
 
+func processWALWithRepair(walDir string, startSegment int, userStates *userStates, nWorkers int,
+	stateCache []map[string]*userState, seriesCache []map[string]map[uint64]*memorySeries, metrics *ingesterMetrics) error {
+
+	corruptErr := processWAL(walDir, startSegment, userStates, nWorkers, stateCache, seriesCache)
+	if corruptErr == nil {
+		return nil
+	}
+
+	metrics.walCorruptionsTotal.Inc()
+	level.Error(util.Logger).Log("msg", "error in replaying from WAL", "err", corruptErr)
+
+	// Attempt repair.
+	level.Info(util.Logger).Log("msg", "attemption repair of the WAL")
+	w, err := wal.New(util.Logger, nil, walDir, true)
+	if err != nil {
+		return err
+	}
+
+	err = w.Repair(corruptErr)
+	if cerr := w.Close(); cerr != nil {
```

Comment:
If both the `w.Repair()` and `w.Close()` fail, we do return the `Close()` error while I believe would be more appropriate to return the `Repair()` error (which is the first one).

Reply:
Right. I will use multi error from TSDB here instead.